### ISSUE TITLE
[FEATURE]: Use Custom Control Buttons as Companion Buttons

### ISSUE THREAD

**Issue by wuffwuff2003** [bitfocus/companion-module-behringer-wing#19]
Hi :)
Is it possible to implement the Custom Controls as Action Buttons (later also the knobs as variables for the new Stream Deck plus). I would do it myself if there was a possibility. I am Beta Tester of the wing, so have the whole docs about the OSC Commands. 
My main goal is to use the TAP Tempo Button, as well as the SendsOnFader with a dedicated BUS (as it is only possible in the CC).
Many thanks

**Comment by istnv** [comment]
Does OSC expose the custom controls?  I was unable to find any way to detect (or press) them on other consoles.

**Comment by wuffwuff2003** [comment]
Hi take a look here 
[WING-OSC-Documentation.pdf](https://github.com/bitfocus/companion-module-behringer-wing/files/11302258/WING-OSC-Documentation.pdf)

**Comment by istnv** [comment]
I have that document. 
Via OSC you can SET the purpose of a button/knob or GET what the button/knob is configured to do.
Unless I am not seeing where, nothing is sent to indicate any particular button/knob is being pushed or rotated.

**Comment by wuffwuff2003** [comment]
Yes I dont understand the programmation of this, but when it is not possible to fire the action behind an CC so I think it would then be possible to fire an Action directly form an effect , like a delay, and then of course this will then result in lots of programmation.

**Comment by wuffwuff2003** [comment]
What I can suggest maybe is when you write the code for one effect, and send me this I can try to add all the current effects in the wing. I only need a presentation of 1 effect. That can help you save time if you wish? Thx :)

**Comment by AntonioPatane** [comment]
I would like the "Disable Bus Assign" to be implemented on the Talkback page, it would be useful in programming direct talk on one bus, excluding the others

**Comment by elliotmatson** [comment]
@janikwitzig I'm looking at this feature. It looks like if you assign any MIDI push action to a button or encoder, it will send updates when pressed or released. The OSC update seems to only be related to the button being pressed, not the MIDI command assigned to it. For example this is pressing user button 1 on a WING Compact:

```
Updating variable: /$ctl/user/U1/1/bu/val [ { type: 'i', value: 127 } ]
Updating variable: /$ctl/user/U1/1/bu/val [ { type: 'i', value: 0 } ]
```

I just messaged in the bitfocus slack to see if there's a common design pattern for triggering companion actions via a module. I feel like the UI is going to be the hard part for this

**Comment by janikwitzig** [comment]
I don't fully understand this request to be honest. Is the idea to eventually be able to control something within companion from the WING?

**Comment by elliotmatson** [comment]
Yea, I think the idea is to be able to use the custom buttons on a Wing as companion buttons. The companion devs are working on a module system for surfaces, and I think we should just wait until that is released for this feature

**Comment by janikwitzig** [comment]
Sounds good to me

**Comment by elliotmatson** [comment]
I'm tempted to resurrect this, as I haven't seen much traffic on surface modules recently.

On the Wing side of things we would have to assign unused MIDI actions as noted above, in order to get OSC updates when pressed. We might want to get the user to pick an entire MIDI channel that is unused, which gives us up to 128 buttons that we could assign at random.

I think it would be ideal if we used the Satellite API to interact with Companion, since that would also allow us to trigger actions as well as get button text and colors. I think initially we can break all of the custom buttons into "pages" similar to how they are shown on the edit dialogues on the Wing, which would leave us with these "surfaces" that could be enabled or disabled in the config:

Full: (18 pages)
  - GPIO (user/gpio): 1x4 (press)
  - User pages 1-16 (user): 3x4  (Top row: rotate, color, text. Bottom Rows: press, text)
  - DAW section user (user/user): 2x4 (press, text)

Compact: (10 pages)
  - GPIO (user/gpio): 1x4 (press)
  - User 1-4 (???): 2x2 (press, color, text)
  - User 5-8 (???): 2x2 (press, color, text)
  - User 9-12 (???): 2x2 (press, color, text)
  - User 13-16 (???): 2x2 (press, color, text)
  - Main/MTX 5-8 (???): 2x2 (press, color, text)
  - DCA 1-4 (???): 2x2 (press, color, text)
  - DCA 5-8 (???): 2x2 (press, color, text)
  - DCA 9-12 (???): 2x2 (press, color, text)
  - DCA 13-16 (???): 2x2 (press, color, text)

Rack: (17 pages)
  - GPIO (user/gpio): 1x4 (press)
  - User pages 1-16 (user): 3x4 (Top row: rotate, color, text. Bottom Rows: press, text)

Maybe this is something I can work on having ready for the major release after the one we are currently preparing for?

**Comment by janikwitzig** [comment]
So for my understanding, you suggest setting up a „remote“ device that then looks like a normal surface to Companion using the Satellite API?

If so, would it potentially make sense to have a separate module for this?

**Comment by elliotmatson** [comment]
A separate module would fit more with what the companion devs are hoping to achieve with surface modules, but then we are back to the problem of the Wing only supporting one OSC subscription. I think these are our options:

1. Wait for companion to have module support for surfaces, and handle forwarding data from the wing module to the wing-surface module
2. Implement Satellite ability in the current module
3. Add variables for the user button status and actions to set the name and color

3 is by far the easiest, we would likely just want to adjust the debouncing of variable updates to be a bit more responsive. 2 feels like the best user experience, but would be a lot of work. 1 feels like the "right" way to do it, but I don't like forwarding data around between modules

**Comment by janikwitzig** [comment]
I played with the Satellite API for a bit. Either there‘s a misunderstanding on my end or what we want to do is not really easy to do.

I originally assumed that when I create a new surface, I‘d have a whole new grid for that surface alone. Companion only maps certain parts of the base grid to a surface. I can create the surface, but from the module‘s point of view I cannot possibly know which buttons to press, because it could be any. This is due to the fact that the API can only press button on page/row/col and not surface/page/row/col

So as I see it, the surface does not solve our issue? Or did you mean to only use the button press actions from the satellite API?

If so, this could be somewhat easy. The user selects a start bank from which the CC banks are mapped, e.g. Bank 81. We then listen for osc events from the CC buttons and press the corresponding ones.

Sorry for being a bit confused about this issue. I‘ll figure it out ;-)

**Comment by janikwitzig** [comment]
You can ignore my previous message, I think I figured it out... played around a little more and I was now finally able to press a single button using the API, still in Python however. This code will probably be easily adaptable to Typescript.

If you want to give it a try, I attached my file below. It starts an interactive shell that allows you to enter the API calls in the console. It creates a single virtual surface named WING_CC. 

[satellite-client.py](https://github.com/user-attachments/files/23145273/satellite-client.py)

## API Connection
A connection can be opened connecting to 'localhost' over port 16622. This is the default Satellite API port. It is specifically mentioned that this should be prepared to be dynamic for possible future uses.

To keep the connection alive, a periodic message "PING <payload>" must be sent. When a "PONG <payload>" is received, this indicates that the connection to the server is alive. If the PING is not sent, the server will terminate the connection. Sending a PING once every 2 seconds is enough.

## Button Press over Satellite API

```
KEY-PRESS DEVICEID=WING_CC KEY=0/1 PRESSED=true
KEY-PRESS DEVICEID=WING_CC KEY=0/1 PRESSED=false
```
on my Satellite surface WING_CC. The socket communicating runs with host 'localhost' over port 16622.

## Implementation Plan
Depending on the desk, we would then create a certain number of Satellite surfaces, each corresponding to a single CC layer. The user can then configure the Companion page to which the CC layer corresponds. In my example, this is page 80.

<img width="1241" height="263" alt="Image" src="https://github.com/user-attachments/assets/84acbe1c-0182-40bc-8359-22eb99ecdbe8" />

The DEVICEID WING_CC is of course not optimal. I think we would rather use something like <custom>_cc_1, where <custom> could be configurable in the module config. This way we can distinguish between multiple consoles controlled from one instance of Companion.

At runtime, we then receive the user button pressed as OSC commands and then automatically know which device id to press, since we fix those when creating the surfaces.

<img width="1619" height="246" alt="Image" src="https://github.com/user-attachments/assets/0706a8a4-4a74-4b92-acd5-13c1cb189cb1" />

## Some Afterthoughts

We would still need a bit of management of the created surfaces. For example, we need to create the surfaces when enabled upon module startup. This can be done using the Satellite API. Since this feature would now finally make use of the Model selection config, we need to also think about how we want to handle when a single instance of the WING companion changes its model type. I tend to explicitly create separate virtual surfaces depending on the connect model; in my experience I tend to have very different working preferences from the full-size console to the compact. This could also be an option toggle, of course.

Then lastly, how would we handle management of those virtual surfaces? For usability, we should not delete them on disconnect or at startup, because if we do, the start page information gets lost and the user would need to re-configure that. But if we do have separate Satellite surfaces per instance of our module, the number of surfaces can get out-of-hand relatively quickly.

**Comment by dmitrylive** [comment]
Hi @janikwitzig! I managed to establish communication between Streamdeck and Wing Compact via RPi + Debian. OSC commands are sent and work, but I also couldn't find any documentation on which commands can emulate button presses in LAYERS 1-16.
Perhaps I misunderstood your idea; mine is that Streamdeck will automatically read the parameters set in the console in user layers LAYERS 1-16. In other words, Streamdeck will essentially be configured from the console.
Basically, if we implement this through the Companion, and there's an option called "....LAYER 1 button 1..... ${realName}," we'll get exactly what we need. And if I understand correctly, we're stumped on which command to use to access these control buttons. Pardon my English, it's not my native language.

**Comment by dmitrylive** [comment]
This is what I was able to find, but I couldn't get it to work.

## Protocol Format
Packet structure (from wcc/libwapi):
- Handshake: 0xDF 0xD1
- Command: 0xD7 + [4-byte hash] + 0xDC

Example: DF D1 D7 36 02 EA D7 DC

## Hash values for User Layer 1 (8 buttons)
$CTL_USER_1_1_BU = 0x3602ead7 (button 0)
$CTL_USER_1_1_BD = 0x00ff3778 (button 1)
$CTL_USER_1_2_BU = 0x1d12d952 (button 2)
$CTL_USER_1_2_BD = 0x480f25f3 (button 3)
$CTL_USER_1_3_BU = 0xa42367ad (button 4)
$CTL_USER_1_3_BD = 0xbf1f53b1 (button 5)
$CTL_USER_1_4_BU = 0x8b335628 (button 6)
$CTL_USER_1_4_BD = 0x2e2f420c (button 7)

## Python send function
```python
def send_button(hash_value):
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.connect(("192.168.0.175", 2222))
    sock.send(struct.pack('>BB', 0xDF, 0xD1))
    sock.send(struct.pack('>BBBBBB', 0xD7,
        (hash_value >> 24) & 0xFF,
        (hash_value >> 16) & 0xFF,
        (hash_value >> 8) & 0xFF,
        hash_value & 0xFF, 0xDC))
    sock.close()

**Comment by janikwitzig** [comment]
Thanks for trying
